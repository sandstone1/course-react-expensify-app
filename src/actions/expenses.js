
//  we will need the uuid import for the AddExpense action generator
import uuid from 'uuid';
// see -- Mark 1 -- below for details
import database from '../firebase/firebase';
// we will need to set up the named exports
// format is " export const action generator name "



// ==============================
// ACTION GENERATOR #1
// ==============================


// we will destructure the first argument and if it doesn't exist we will destructure
// an empty object
// we can define all the things we want within the first {} within the argument
// we want the user to provide the value for description and note but if the user
// does not provide a value then an empty string will surfice
// next up, we default the amount to an appropriate value like 0
// next up, we default the createdAt to an appropriate value like 0 until we talk about
// timestamps
export const addExpense = ( expense ) => ( {
    type : 'ADD_EXPENSE',
    expense : expense
} );



// ==============================
// ACTION GENERATOR #2
// ==============================


// we need the id from the expense we are going to remove
export const removeExpense = ( { id } = {} ) => ( {
    type : 'REMOVE_EXPENSE',
    expense : {
        id : id
    }
} );




// ==============================
// ACTION GENERATOR #3
// ==============================


// pass in our 2 arguments: id and updates and neither need default values since we
// we don't have the ID and were not updating anything ( don't understand this last
// part ) 
export const editExpense = ( id, updates ) => ( {
    type : 'EDIT_EXPENSE',
    id, // remember this is equal to id : id
    updates // remember this is equal to updates : updates
} );


// -- Mark 1 --
// lecture 152: Asynchronous Redux Actions
// addExpense will dispatch an object and that object will change the redux store whereas
// startAddExpense is just going to start that process off and startAddExpense will dispatch
// addExpense inside the function below and that dispatch will change the redux store and in
// the past we always returned objects but in this case we are going to return a function and
// this function works because we set up the middleware for redux-thunk and by default doing
// something like this would not work

// the return function below gets called internally by redux and it gets called with dispatch
// as the argument and this gives us access to dispatch inside our return function and we
// use dispatch after we write some data to Firebase and after waiting for that data to correctly
// sync and then after that we will use dispatch to dispatch addExpense and we will make sure the
// redux store reflects those changes as well and if we going to be saving to Firebase inside the
// return function we will need to restructure addExpense above so instead of having the
// default in the addExpense action generator we want to put the default key value pairs inside
// the return function and will switch up how we set those defaults in the argument to
// startAddExpense

// we will call the argument to startAddExpense " expenseData " and by default set it equal to an
// empty object and set up a const inside the return function and destructure a bunch of stuff and
// we'll destructure from expenseData and then copy the addExpense argument and place it inside the
// const object below and this is exactly the same argument as we had in the addExpense action
// generator and since we are setting up the defaults below that means addExpense can get a whole
// lot simplier

// the next thing we are going to do is use push to save some data to Firebase and we first need
// to import the database variable into this file so we will do this above:
// " import database from '../firebase/firebase'; " and then we can use the database variable below
// and we will get a reference to " expenses " and then we will push an expense, which is an object,
// into our Firebase database and we will create an variable for this object
// and call it expense and then pass that variable into push()
// and then we can attach the then call and at this point what is going to happen? the function
// will run and then we will take all the data or expenseData and save the data to Firebase and at
// this point there is still an important step left and that step is to dispatch the addExpense
// action generator from above and change the addExpense action generator above to this:
/*
export const addExpense = ( expense ) => ( {
    type : 'ADD_EXPENSE',
    expense : expense
} );
*/

// and yes we broke some test cases and we will fix those in the next video but for now let's focus
// on getting this functionality working and what are we trying to add as the argument to the
// addExpense call below? we are trying to add the expense we just created in the expense variable
// up above and we have to include the ID as well in the addExpense call and the good news for us
// is that the then callback or success case gets called with a reference and we can access that
// reference by including the reference as the argument to then() and we'll call the reference
// " ref " and we can access the ref and do something meaningful with it and in our case we'll use
// ref to get the id by creating an object inside the addExpense call and defining the ID as
// follows: " id : ref.key " and then we'll tack on all the properties in expense by using
// the spread operator or " ...expense "
export const startAddExpense = ( expenseData = {} ) => {

    return ( dispatch ) => {

        const {
            description = '',
            note        = '',
            amount      = 0,
            createdAt   = 0
        } = expenseData;

        const expense = { description, note, amount, createdAt };

        // add return below and see -- Mark 1 -- of lecture 153: Testing Async Redux Actions:
        // Part 1 for details on why we added the return keyword below
        return database.ref( 'expenses' ).push( expense ).then( ( ref ) => {
            
            dispatch( addExpense( { 
                id : ref.key,
                ...expense
            } ) );
        
        } );

    };

};

// and there we have it, this is all we need to do inside expenses.js and the last step before this
// is going to work is to make sure we dispatch startAddExpense as opposed to dispatching
// addExpense and

// -- go to the AddExpensePage component page --

// and change:
// " import { addExpense } from '../actions/expenses'; " to
// " import { startAddExpense } from '../actions/expenses'; ""

// and change:
// " addExpense : ( expense ) => dispatch( addExpense( expense ) ) " to
// " startAddExpense : ( expense ) => dispatch( startAddExpense( expense ) ) "

// and change:
// " this.props.addExpense( expense ); " to 
// " this.props.startAddExpense( expense ); "

// with these changes in place we are all done and we can test this out and test this out by
// creating an expense and what we hope to see is the expense showing up in the browser meaning it
// was added to redux and we are also hoping the expense shows up in our Firebase database and
// after I add an expense in the expensify site I get redirected to the home page and I see my
// expense listed meaning it was added to the redux store and when I go to Firebase I see our
// new expense listed in the database

// so by using redux-thunk were going to be able to create asynchronous actions and these actions
// will do something first soemthing asynchronous like make a Firebase data call and then these
// actions will use dispatch to actually change the redux store

// remember, this is just one of our 4 CRUD operations so we have create done but were still not
// reading from Firebase but if we do a page refresh we lose the data and this is expected for the
// moment and we'll read from Firebase in just a little bit but what I want to focus on in the
// next video though is how we can test the code we just wrote in this video and we want to make
// sure that everything we expect to happen actually happens and we will change some test cases
// for things we changed in this video and remember we changed the addExpense action generator so
// we have some breaking test cases there and we changed the AddExpensePage component so we will
// need to make some changes to that test case as well


// -- END OF MARK 1 --
