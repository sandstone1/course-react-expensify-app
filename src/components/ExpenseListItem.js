
// -- Mark 1 --

// challange
// (1) we want to export a stateless functional component
// and this component will render a few pieces of information
// about the individual expense and those pieces are: (a) description
// (b) amount and (c) createAt value

// import react since we are using some jsx
import React from 'react';
// -- Mark 2 --
// -- Mark 4 -- from EditExpensePage
// comment out the connect and removeExpense named exports below
// ------------
// import {       connect } from         'react-redux';
// import { removeExpense } from '../actions/expenses';
// -----------

// ==============================
// GO TO EDITEXPENSEPAGE.JS -- Mark 7 --
// ==============================

// -- Mark 3 --
// lecture 108: Wiring up Edit Expense
// import the named export Link
import {          Link } from 'react-router-dom';
// -----------

// -- Mark 5 --
import moment from 'moment';
import numeral from 'numeral';
// ------------



// we will implicitly return some jsx
// we are going to destructure the props object or props.expense and get the individual
// values off of it
// this works     => const ExpenseListItem = ( { description, amount, createdAt } = props )
// this works too => const ExpenseListItem = ( { description, amount, createdAt } = props.expense )
const ExpenseListItem = ( { dispatch, id, description, amount, createdAt } ) => (
    
    // -- Mark 2 -- 
    // from ExpenseListFilters.js
    // challange lecture 103
    // create a remove list item button below
    // the remove button will dispatch an action when it get clicked
    // so we will (1) need to import the correct action generator from the actions/expenses.js
    // file and (2) we are going to have to connect this component in order to access dispatch and
    // (3) we going to have to wire up onClick to get everything working correctly

    // in the argument above, we continue to use destructuring and grab dispatch and id
    // remember, all removeExpense needs is the id and id was passed in above as one of the
    // arguments
    // remember, we set up the removeExpense action generator to take an object with the id
    // value inside the object so we need to set this up below in order to make our action
    // generator work properly
    // we could do { id : id } or just use the ES6 shorthand way ( i.e. { id } )

    // lecture 103 recap
    // more of our components are communicating with the redux store, which us great
    // we are going to continue going through this process focusing on the filters in
    // the next video

    // -- Mark 3 --
    // lecture 108: Wiring up Edit Expense
    // need to template string inside the Link and interpolate the id variable by typing ${}

    // ==============================
    // GO TO EDITEXPENSEPAGE.JS -- Mark 4 --
    // ==============================

    // ------------

    // -- Mark 4 -- from EditExpensePage
    // challange
    // remove the button below and copy it into the EditExpensePage component and remove the
    // related imports above and the connect function below since we no longer need the connect
    // function
    // -----------

    // -- Mark 5 --
    // lecture 138: New Feature Workflow
    // move amount and createdAt onto their own lines
    // first, we're going to focus on formatting createdAt with moment and need to import moment
    // above and below we need to change { createdAt } to { moment(createdAt).format( 'MMMM Do, YYYY' ) }
    // and this will create a moment instance with our timestamp as the argument and then we add
    // in the format method with the follwoing string: 'MMMM Do, YYYY'

    // next up, we need to format amount and were going to be using a library that will make it really
    // easy to format currency and the library's name is numeral.js and we find information on this
    // library at numeraljs.com and the website says that Numeral.js is a " A javascript library for
    // formatting and manipulating numbers. " so below we will need to change { amount } to
    // { numeral( amount / 100 ).format( '$0,0.00' ) } and we had to devide amount by 100 to convert
    // cents to dollars
    // and we will call numeral and pass out amount in as the argument which will give us our numeral
    // instance and then we will call the format method with the pattern we want to use

    // now go to the expensify site and make sure the amount and time are showing correctly and
    // now that everything is working correctly, we can shut down the dev-server and make a commit
    // but first run 
    
    // " git status "
    
    // to get an overview of what we've changed and then we can make
    // the commit by typing:
    
    // " git commit -a -m "Set up formatting for amount and createdAt" "
    
    // and can go straight to making a commit since we only have files with modified changed and no
    // untracked files and then we can push our changes to github by typing:

    // " git push "

    // and once our changes are in our github remote repository then we can push our changes to Heroku
    // and deploy our code on the web and do this we type:

    // " git push heroku master "

    // and once the code is up on the Heroku servers Heroku will incorporate our latest changes and
    // redeploy the application
    <div>
        <Link to={ `/edit/${ id }` }>
            <h3>{ description }</h3>
        </Link>
        <p>
            { numeral( amount / 100 ).format( '$0,0.00' ) }
            -
            { moment(createdAt).format( 'MMMM Do, YYYY' ) }
        </p>
    </div>
    // -----------
);

// could have done it this way instead of the way above
// ** way #2 **
/*
const ExpenseListItem = ( props ) => (
    <div>
        <h3>{ props.expense.description }</h3>
        <p>{ props.expense.amount } - { props.expense.createdAt }</p>
    </div>
);
*/


// -- Mark 2 --

// export default ExpenseListItem;
// change " export default ExpenseListItem; " to
// " export default connect()( ExpenseListItem ) "
// we are going to export the connected component
// for this challange, we don't need anything from the store we just want connect
// the ExpenseListItem component so that we are able to access the dispatch prop

// -- Mark 4 -- from EditExpensePage
// remove the connect function
// ------------
export default ExpenseListItem

// -----------


// ==============================
// GO TO EXPENSELIST.JS -- Mark 5 --
// ==============================