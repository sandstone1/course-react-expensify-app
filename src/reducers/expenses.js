

// there are no dependencies below and we are not using any sort of library
// so there is no need to import anything and we are not using jsx below so
// there is no need to import react



// EXPENSES REDUCER
// pass in the 2 arguments: state and action and return the new state
// the default state value will be an empty array and we can create a
// variable for our default state value
const expensesReducerDefaultState = [];

// USE A DEFAULT EXPORT BELOW
// we will use a default export since there is one thing we want to export from
// this file
// for the default export, as the first line below, we could do:
// " const expensesReducer = ( state = expensesReducerDefaultState, action ) => { "
// and then " export default expensesReducer; "
// or we do
// " export default ( state = expensesReducerDefaultState, action ) => { "

export default ( state = expensesReducerDefaultState, action ) => {
    // we will switch what we do based off a particular value
    switch ( action.type ) {
        // inside the {} we can define the various cases we want to handle
        case 'ADD_EXPENSE':
            // here we want to run some code that adds the expense onto the array
            // don't want to use state.push( action.expense ) since push will change
            // the original array and we want to avoid this so we will use
            // state.concat( action.expense ) since this takes the state array and combines
            // it with action.expense to create a new array and we return this array
            /*
            return state.concat( action.expense );
            */
            
            // SPREAD OPERATOR
            // now the above works but we want to explore the ES6 spread operator and
            // this allows us to get the same thing done and more
            // as an example, const names = [ 'Andrew', 'Steve' ];
            // if I use the spread operator or [ ...names  ], the result is [ 'Andrew', 'Steve' ]
            // if I want to add an item onto the names array and return a new array, I do as
            // follows [ ...names, 'Mike' ] and the result is [ 'Andrew', 'Steve', 'Mike' ]
            // the great thing here is the spread operator did not change the names array
            // could do [ 'Adam', ...names, 'Mike' ] and the result is [ 'Adam', Andrew', 'Steve', 'Mike' ]
            // so instead of doing this " return state.concat( action.expense ); ", were going to do
            // return [ ...state, action.expense ]; ( so were taking the current state array and adding
            // on the action.expense object without changing the state array )
            return [ ...state, action.expense ];

        case 'REMOVE_EXPENSE':
            // use filter to remove an expense from the state array
            // filter creates a brand new array from another array and does this by looking at each item in the array
            // and will call a function once for each item in the array and the function will return true or false
            // if the function returns true then the current item will be passed into the new array
            // if the function returns false then the current item will not be passed into the new array

            // remember, action.expense.id is equal to expenseOne.expense.id
            // could use destructuring here so { id } = item (remember we have an object here) so we can pull
            // id off each item in the state array
            // BELOW WORKS
            // return state.filter( ( { id } ) => id !== action.expense.id )
            // BELOW WORKS
            return state.filter( ( item ) => item.id !== action.expense.id )

        case 'EDIT_EXPENSE':
            // will use the object spread operator below
            // what do we want to do here? we want to go through every singe expense in the array and
            // find the match and when we find the match then go ahead and correctly change the match
            // we going to use a some conditional logic to change the expense when our item.id matches
            // the action.id
            return state.map( ( item ) => {
                if ( item.id === action.id ) {
                    // if we have a match, we want to return a brand new object
                    return {
                        // will use the object spread operator below
                        // we will spread out the expense item so we want to grab description,
                        // note, amount and createdAt and we just want to override the previous values
                        // with the ones that the user passed in
                        // ...item will give us current item / expense key value pairs and ..action.updates
                        // will provide new key value pairs which will override the previous key value
                        // pairs and this will result in new key value pairs for this item / expense
                        // this results in a new key value pair or " amount : 500 "
                        ...item,
                        ...action.updates

                        // by using the object spread operator we have simple way to create a new object
                        // from an existing oject and we will be using the object spread operator
                        // extensively throughout the course
                    }
                }
                else {
                    // do nothing
                    return item;
                }
            } );

        // -- Mark 1 --
        // lecture 157: Fetching Expenses: Part 1
        case 'SET_EXPENSES':

            // all we need to do is return action.expenses and we are accounting for whatever
            // is in the state and we do not care about the previous expenses and this case
            // is designed to set the expenses array completely and now let's go to the
            // tests/reducers/expenses.test.js file and set the test case
            return action.expenses;

        // End of -- Mark 1 --


        // set up the default case
        default :
            // if the store is running for the first time, then the default state value
            // will be shown
            return state;
    }
};
