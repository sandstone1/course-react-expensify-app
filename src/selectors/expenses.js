
// -- Mark 1 --
// lecture 110: Filtering by Dates
import moment from 'moment';
// -----------

// there are no dependencies below and we are not using any sort of library
// so there is no need to import anything and we are not using jsx below so
// there is no need to import react


// get visible expenses
// will take 2 arguments: the expenses array and the filters object and using
// these 2 pieces of data, we can calculate what the visible expenses should
// look like

// startDate and endDate are timestamps and are in milliseconds
// timestamps start at January 1, 1970 ( known as the unix epoch )
// we will be using timestamps in this course as a way to store
// timezone independent time data

// USE A DEFAULT EXPORT BELOW
// we will use a default export since there is one thing we want to export from
// this file
// for the default export, as the first line below, we could do:
// " const getVisibleExpenses = ( expenses, { text, sortBy, startDate, endDate } ) => { "
// and then " export default getVisibleExpenses; "
// or we could do
// " export default ( expenses, { text, sortBy, startDate, endDate } ) => { "

// change filters to text, sortBy, startDate, endDate
export default ( expenses, { text, sortBy, startDate, endDate } ) => {
    // use filter to
    // filter creates a brand new array from another array and does this by looking at
    // each item in the array and will call a function once for each item in the array
    // and the function will return true or false
    // if the function returns true then the current item will be passed into the new array
    // if the function returns false then the current item will not be passed into the new array
    return expenses.filter( ( expense ) => {
        // we will make 3 const to store whether or not we have a match
        // for text, if I enter in some search text I want to show the expenses that actually
        // use that text value somewhere
        // for our purposes, we are just going to check the description and not the notes

        // challange for textMatch
        // 1 -- figure out if expense.description is inside the text variable string
        // 2 -- use the includes method and convert both strings to lower case
        const textMatch      = expense.description.toLowerCase().includes( text.toLowerCase() );
    
        // -- Mark 1 --
        // lecture 110: Filtering by Dates
        // first, we need to figure how to make the comparisons below using moment
        // and to do this we are going to be using the Query methods from moment
        // comment out the following startDateMatch and endDateMatch
        // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
        // const endDateMatch   = typeof endDate   !== 'number' || expense.createdAt <= endDate;
        // startDateMatch will always be true if there is no startDate because we will never filter
        // out an expense when there is no startDate or the same goes for endDate
        // create a new moment to use as the argument to isSameOrBefore()
        // we passed in " day " as the second argument and therefore we will use day as the unit
        // of measurement so is startDate the same day or a day or days before the createdAtMoment
        const createdAtMoment = moment( expense.createdAt );
        const startDateMatch = startDate ? startDate.isSameOrBefore( createdAtMoment, 'day' ) : true ;
        const endDateMatch   = endDate ? endDate.isSameOrAfter( createdAtMoment, 'day' ) : true;
        // if all three const matches from above are true then we have a complete match and we want
        // to return true from filter and in this case the expense will be keeped ( or be passed into
        // the new array of expenses )
        // if we do not have a complete match then the expense will go away

        // so by setting up the DateRangePicker were going to be able to filter our expenses via this
        // calender widget and that's going to make searching and navigating our expenses much easier
        // we have 2 ways to filter: by date or description and we have 1 way to sort: by date or amount
        // recap: we changed startDate and endDate default state values to moments in time in the filter
        // reducer and we added the RangeDatePicker in the ExpenseListFilters component and the last thing
        // we changed we how we compared the values in startDateMatch and endDateMatch from above
        // ------------
        return textMatch && startDateMatch && endDateMatch;

    // now it's time to add sorting so we can get visible expenses
    // we want to be able to sort by date or sort by amount

    // we are going to look at 2 expenses and determine which one should come
    // before the other

    // sort gets called on an array and returns an array
    // when need to write a compare function or arguments ( a, b ) and then
    // compare them to determine which one comes first
    } ).sort( ( a, b ) => {
        // first thing we need to figure out are we sorting by date or amount
        if ( sortBy === 'date'  ) {
            // if we return -1 a would come first or if we return 1 b would come first
            // if a.createdAt < b.createdAt then we will show b first or have the
            // most recent expenses at the top of the list
            // use the ternary operator and if a.createdAt < b.createdAt then the
            // truthy value will be 1 ( b comes first ) otherwise a comes first
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        // challange
        // sort by amount and put the one with a greater amount first
        else if ( sortBy === 'amount' ) {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};
