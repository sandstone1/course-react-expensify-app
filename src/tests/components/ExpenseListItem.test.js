
// ==============================
// COMING FROM TESTS/COMPONENTS/EXPENSELIST.TEST.JS
// ==============================

// -- Mark 1 --
// lecture 120: Snapshot Testing with Dynamic Components
// first, let's import the things we need
// and will need react since we are going to be using a little bit of jsx inside this file
import React from 'react';
// we will shallow render our component
import { shallow } from 'enzyme';
// import the ExpenseListItem component and remember this is a default export
import ExpenseListItem from '../../components/ExpenseListItem';
// we are going to be using the expenses array from fixtures/expenses.js so need to import
// the dummy expenses data
import expenses from '../fixtures/expenses';

// these are the imports we need to snapshot the ExpenseListItem component
// let's create a new test
// so were going to see what happens when we render the ExpenseListItem component with the given
// expenses
// TEST CASE #1
test( 'should render ExpenseListItem with expense item #2', () => {
    // now were going to render the component with that data and make an assertion about it
    // using the snapshot functionality
    // so create a const called wrapper and we are going to get the wrapper value from whatever
    // comes back from calling shallow() and we are going to call shallow with the jsx that we
    // would like to render and in this case, we are going to render an instance of the
    // ExpenseListItem component and this component requires the id, description, amount and
    // createdAt values from one expense item, which in our case will be the second expense item
    // from the fixtures/expenses.js file

    // by destructuring expense item #2, we are able to get the 4 required values off of the
    // object
    const wrapper = shallow( <ExpenseListItem { ...expenses[ 1 ] } /> );

    // now that we have the actual rendered component, what we need to do is set up a snapshot
    // so I am going to expect something about the wrapper and I will expect the wrapper to
    // match the snapshot and remember the first time we run this test case there is no existing
    // snapshot so the test can never actually fail and will go through the process of creating
    // a new snapshot file for this test file
    expect( wrapper ).toMatchSnapshot();
} );

// now if we ever make any changes to the component we will be alerted because we now have a
// snapshot tracking this component

// there are 2 more components we want to write snapshot tests for and those 2 components are:
// the ExpenseDashBoardPage component and the NotFoundPage component and we will be adding
// snapahots for both of these before the end of the video and these snapshots will be very similar
// to our Header component snapshot and it is a very simple setup, we just render the component with
// no props and we make a snapshot assertion about it so the instructor wants me to create test
// cases for both of these components and again the 4 steps are:
// step 1: create a test file and
// step 2: grab the necessary imports and
// step 3: we will create a test case that renders the actual component and we will render
// both components with no props
// step 4: create the snapshot
// and then go into the snapshots folder and make sure both snapshots look good 


// ==============================
// GO TO TESTS/COMPONENTS/EXPENSEDASHBOARDPAGE.TEST.JS -- MARK 1 --
// ==============================
