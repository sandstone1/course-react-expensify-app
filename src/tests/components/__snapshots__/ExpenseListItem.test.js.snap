// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expense item #2 1`] = `
<div>
  <Link
    to="/edit/2"
  >
    <h3>
      Rent
    </h3>
  </Link>
  <p>
    $1,095.00
    -
    December 27th, 1969
  </p>
</div>
`;
